### PR TITLE
OAuth token refreshes bug fix + optimization

### PR DESCRIPTION
**Pull request by tasherif-msft**
Currently any large job using OAuth will eventually hit a service error we're refreshes are happening too frequently, due to the nature of jobs concurrently running in parts, the token is refreshed too many times rather than reusing the already created fresh token. This PR fixes that - it also led to better perf due to the lower number of service calls.

Resolves https://github.com/Azure/azure-storage-azcopy/issues/1984
Resolves https://github.com/Azure/azure-storage-azcopy/issues/1957

### CODE REVIEW

File: ste/mgr-JobMgr.go [hunk 1]
```diff
@@ -712,13 +713,15 @@ func (jm *jobMgr) CloseLog() {
 
 // DeferredCleanupJobMgr cleanup all the jobMgr resources.
 // Warning: DeferredCleanupJobMgr should be called from JobMgrCleanup().
-//          As this function neither threadsafe nor idempotient. So if DeferredCleanupJobMgr called
-//          mulitple times, it may stuck as receiving channel already closed. Where as JobMgrCleanup()
-//          safe in that sense it will do the cleanup only once.
+//
```
Comment (gapra-msft) [inline ste/mgr-JobMgr.go:716]: nit Why are there new lines of // added here?

File: ste/mgr-JobMgr.go [hunk 2]
```diff
@@ -961,7 +964,14 @@ func (jm *jobMgr) scheduleJobParts() {
 				go jm.poolSizer()
 				startedPoolSizer = true
 			}
-			jobPart.ScheduleTransfers(jm.Context())
+			getCredential := true
```
Comment (gapra-msft) [inline ste/mgr-JobMgr.go:967]: nit getCredential := jm.credential == nil

File: ste/mgr-JobMgr.go [hunk 3]
```diff
@@ -961,7 +964,14 @@ func (jm *jobMgr) scheduleJobParts() {
...
+			if jm.credential != nil {
+				getCredential = false
+			}
+			jobPart.ScheduleTransfers(jm.Context(), getCredential, jm.credential)
```
Comment (gapra-msft) [inline ste/mgr-JobMgr.go:971]: I'm confused here, shouldnt we get the credential before scheduling transfers?

Comment (tasherif-msft) [reply ste/mgr-JobMgr.go:971]: for minimal code changes in the first job part (and its corresponding scheduleTransfers call) we store that credential retrieved when the pipeline is created which is then used for other job parts.

Comment (gapra-msft) [reply ste/mgr-JobMgr.go:971]: So basically JobMgr and JobParts will have separate creds?

Comment (tasherif-msft) [reply ste/mgr-JobMgr.go:971]: JobMgr will store the cred after the first jobPart does its first scheduleTransfers call and then pass it to the next jobParts

File: ste/mgr-JobMgr.go (hunk 2 above)
Comment (gapra-msft) [inline ste/mgr-JobMgr.go:967]: also nit: should we call this getSourceCredential? I know in AzCopy they seem to be very explicit about source vs dest names

File: ste/mgr-JobMgr.go (hunk 1 above)
Comment (tasherif-msft) [reply ste/mgr-JobMgr.go:716]: It looks like GO is formatting this automatically

File: ste/mgr-JobMgr.go [hunk 4]
```diff
@@ -961,7 +964,11 @@ func (jm *jobMgr) scheduleJobParts() {
 				go jm.poolSizer()
 				startedPoolSizer = true
 			}
-			jobPart.ScheduleTransfers(jm.Context())
+			getSourceCredential := jm.credential == nil
+			jobPart.ScheduleTransfers(jm.Context(), getSourceCredential, jm.credential)
```
Comment (vibhansa-msft) [inline ste/mgr-JobMgr.go:968]: getSourceCredential is just a flag saying jm.credential is nil or not. Here we are passing down both of parameters here. Flag can be ignored and directly jm.credential can be checked internally?

File: ste/mgr-JobPartMgr.go [hunk 5]
```diff
@@ -561,9 +561,13 @@ func (jpm *jobPartMgr) createPipelines(ctx context.Context) {
 				CallerID: fmt.Sprintf("JobID=%v, Part#=%d", jpm.Plan().JobID, jpm.Plan().PartNum),
 				Cancel:   jpm.jobMgr.Cancel,
 			}
-
-			sourceCred = common.CreateBlobCredential(ctx, jobState.CredentialInfo.WithType(jobState.S2SSourceCredentialType), credOption)
```
Comment (vibhansa-msft) [inline ste/mgr-JobPartMgr.go:565]: So, actual fix here is to reuse the same credential instead of creating a new one for each pipeline?

**Comment by vibhansa-msft** [comment]
How are we fixing the refreshToken? If multiple goroutine detect token is expired they can still go in parallel to refresh the token,

File: ste/mgr-JobPartMgr.go (hunk 5 above)
Comment (tasherif-msft) [reply ste/mgr-JobPartMgr.go:565]: yes

**Comment by tasherif-msft** [comment]
> How are we fixing the refreshToken? If multiple goroutine detect token is expired they can still go in parallel to refresh the token,

The issue that causes the job to fail isn't that multiple goroutines are hitting refresh simultaneously, its simply that each job is refreshing the token. When I ran this multiple times on a large job it never hit really hit the case you're referencing. So, I decided to keep changes to a minimum and focus on fixing reusing the token. I'd rather not overengineer here.

File: ste/mgr-JobMgr.go (hunk 4 above)
Comment (tasherif-msft) [reply ste/mgr-JobMgr.go]: done

**Pull request closed without merging**